README.md
**This is a version of the GHUnit framework containing custom modifications made by NumberFour AG. The original source can be found at gabriel/gh-unit**
***

# GHUnit

GHUnit is a test framework for Objective-C (Mac OS X 10.5 and above and iPhone 3.x and above).
It can be used with SenTestingKit, GTM or all by itself. 

For example, your test cases will be run if they subclass any of the following:

- GHTestCase
- SenTestCase
- GTMTestCase

## Download

[See Downloads](http://github.com/gabriel/gh-unit/downloads)

## Why?

The goals of GHUnit are:

- Runs unit tests, allowing you to breakpoint and interact with the XCode Debugger.
- Ability to run from Makefile's or the command line.
- Run tests in parallel.
- Allow testing of UI components.
- View metrics; Search and filter tests; View logging by test case.
- Show stack traces and useful debugging information.
- Be embeddable as a framework (using @rpath) for Mac OSX apps, or as a static library in your iPhone projects.

Future plans include:

- Mocks (maybe integrate OCMock?)

`GHTestCase` is the base class for your tests.

- Tests are defined by methods that start with `test`, take no arguments and return void. For example, `- (void)testFoo { }`
- Your setup and tear down methods are `- (void)setUp;` and `- (void)tearDown;`. 
- Your class setup and tear down methods are `- (void)setUpClass;` and `- (void)tearDownClass;`. 
- By default tests are run on a separate thread. For a UI test or to run on the main thread, implement:

    `- (BOOL)shouldRunOnMainThread { return YES; }`

## Group

For questions and discussions see the [GHUnit Google Group](http://groups.google.com/group/ghunit)

## Adding a GHUnit Test Target (Mac OS X)

There are two options. You can install it globally in /Library/Frameworks or with a little extra effort embed it with your project.

### Installing in /Library/Frameworks

- Copy `GHUnit.framework` to `/Library/Frameworks/`
- Add a `New Target`. Select `Cocoa -> Application`. Name it `Tests` (or something similar).
- In the `Target 'Tests' Info` window, `General` tab:
	- Add a linked library, under `Mac OS X 10.X SDK` section, select `GHUnit.framework`
	- If your main target is a library: Add a linked library, and select your main target.
	- If your main target is an application, you will need to include these source files in the `Test` project manually. 
	- Add a direct dependency, and select your project. (This will cause your application or framework to build before the test target.)
- Copy [GHUnitTestMain.m](http://github.com/gabriel/gh-unit/tree/master/Classes-MacOSX/GHUnitTestMain.m) into your project and include in the Test target.
- Now create a test (either by subclassing `SenTestCase` or `GHTestCase`), adding it to your test target. (See example test case below.)
- By default, the Tests-Info.plist file includes `MainWindow` for `Main nib file base name`. You should clear this field.
- (Optional) Install Makefile (see instructions below)

### Installing in your project

- Add a `New Target`. Select `Cocoa -> Application`. Name it `Tests` (or something similar).
- In the Finder, copy `GHUnit.framework` to your project directory (maybe in MyProject/Frameworks/.)
- In the `Tests` target, add the `GHUnit.framework` files (from MyProject/Frameworks/). It should now be visible as a `Linked Framework` in the target. 
- In the `Tests` target, under Build Settings, add `@loader_path/../Frameworks` to `Runpath Search Paths` (Under All Configurations)
- In the `Tests` target, add `New Build Phase` | `New Copy Files Build Phase`. 
	- Change the Destination to `Frameworks`.
	- Drag `GHUnit.framework` into the the build phase
	- Make sure the copy phase appears before any `Run Script` phases 
- Copy [GHUnitTestMain.m](http://github.com/gabriel/gh-unit/tree/master/Classes-MacOSX/GHUnitTestMain.m) into your project and include in the Test target.

- If your main target is a library: 
	- In the `Target 'Tests' Info` window, `General` tab: 
		- Add a linked library, and select your main target; This is so you can link your test target against your main target, and then you don't have to manually include source files in both targets.
- If your main target is an application, you will need to include these source files to the `Test` project manually.

- Now create a test (either by subclassing `SenTestCase` or `GHTestCase`), adding it to your test target. (See example test case below.)
- By default, the Tests-Info.plist file includes `MainWindow` for `Main nib file base name`. You should clear this field.
- (Optional) Install Makefile (see instructions below)

### Example test case (Mac OS X)

For example `MyTest.m`:

	#import <GHUnit/GHUnit.h>

	@interface MyTest : GHTestCase { }
	@end

	@implementation MyTest
	
	- (BOOL)shouldRunOnMainThread {
		// By default NO, but if you have a UI test or test dependent on running on the main thread return YES
	}
	
	- (void)setUpClass {
		// Run at start of all tests in the class
	}

	- (void)tearDownClass {
		// Run at end of all tests in the class
	}

	- (void)setUp {
		// Run before each test method
	}

	- (void)tearDown {
		// Run after each test method
	}

	- (void)testFoo {		
		GHTestLog(@"I can log to the GHUnit test console: %@", foo);
		
		// Assert a is not NULL, with no custom error description
		GHAssertNotNULL(a, nil);
	
		// Assert equal objects, add custom error description
		GHAssertEqualObjects(a, b, @"Foo should be equal to: %@. Something bad happened", bar);
	}

	- (void)testBar {
		// Another test
	}

	@end

Now you should be ready to Build and Run the test target.

You should see something like:

![GHUnit-0.4.18](http://rel.me.s3.amazonaws.com/gh-unit/images/GHUnit-0.4.18.png)

- Optionally, you can create and and set a prefix header (`Tests_Prefix.pch`) and add `#import <GHUnit/GHUnit.h>` to it, and then you won't have to include that import for every test.

## Adding a GHUnit Test Target (iOS)

- Add a `New Target`. Select `Cocoa Touch -> Application`. Name it `Tests` (or something similar).
- Add the `GHUnitIOS.framework` to your project.
- Add the following frameworks to `Linked Libraries`:
  - `GHUnitIOS.framework`
  - `CoreGraphics.framework`
  - `Foundation.framework`
  - `UIKit.framework`
  - (Optional) `CoreLocation.framework`
- Under 'Framework Search Paths' make sure the (parent) directory to GHUnitIOS.framework is listed.
- Under 'Other Linker Flags' in the `Test` target, add `-ObjC` and `-all_load`
- By default, the Tests-Info.plist file includes `MainWindow` for `Main nib file base name`. You should clear this field.
- Add the [GHUnitIOSTestMain.m](http://github.com/gabriel/gh-unit/blob/master/Project-IPhone/GHUnitIOSTestMain.m) file into your project.
- (Optional) Install Makefile (see instructions below)
- (Optional) Create and and set a prefix header (`Tests_Prefix.pch`) and add `#import <GHUnitIOS/GHUnitIOS.h>` to it, and then you won't have to include that import for every test.

Now you can create a test (either by subclassing `SenTestCase` or `GHTestCase`), adding it to your test target.


### Example test case (iPhone)

For example `MyTest.m`:

	#import <GHUnitIOS/GHUnitIOS.h>

	@interface MyTest : GHTestCase { }
	@end

	@implementation MyTest

	- (BOOL)shouldRunOnMainThread {
		// By default NO, but if you have a UI test or test dependent on running on the main thread return YES
	}
	
	- (void)setUpClass {
		// Run at start of all tests in the class
	}

	- (void)tearDownClass {
		// Run at end of all tests in the class
	}
	
	- (void)setUp {
		// Run before each test method
	}

	- (void)tearDown {
		// Run after each test method
	}	

	- (void)testFoo {
		// Assert a is not NULL, with no custom error description
		GHAssertNotNULL(a, nil);
	
		// Assert equal objects, add custom error description
		GHAssertEqualObjects(a, b, @"Foo should be equal to: %@. Something bad happened", bar);
	}

	- (void)testBar {
		// Another test
	}

	@end

Now you should be ready to Build and Run the `Test` target.

You should see something like:

![GHUnit-IPhone-0.4.18](http://rel.me.s3.amazonaws.com/gh-unit/images/GHUnit-IPhone-0.4.18.png)

- Optionally, you can create and and set a prefix header (`Tests_Prefix.pch`) and add `#import "GHUnit.h"` to it, and then you won't have to include that import for every test.

An example of an iPhone project with GHUnit test setup can be found at: [MyTestable-IPhone](http://github.com/gabriel/gh-unit/tree/master/Examples/MyTestable-IPhone).

## GHUnit Environment Variables

Go into the "Get Info" contextual menu of your (Tests) executable (inside the "Executables" group in the left panel of XCode). 
Then go in the "Arguments" tab. You can add the following environment variables:

	GHUNIT_CLI - Default NO; Runs tests on the command line (see Debugger Console, Cmd-Shift-R)
	GHUNIT_RERAISE - Default NO; If an exception is encountered it re-raises it allowing you to crash into the debugger
	GHUNIT_AUTORUN - Default NO; If YES, tests will start automatically
	GHUNIT_AUTOEXIT - Default NO; If YES, will exit upon test completion (no matter what); For command line MacOSX testing

## Test Environment Variables (Recommended)

Go into the "Get Info" contextual menu of your (Tests) executable (inside the "Executables" group in the left panel of XCode). 
Then go in the "Arguments" tab. You can add the following environment variables:
	 

	Environment Variable:                 Default:  Set to:
	NSDebugEnabled                           NO       YES
	NSZombieEnabled	                         NO       YES
	NSDeallocateZombies                      NO       NO (or YES)
	NSHangOnUncaughtException                NO       YES
	NSAutoreleaseFreedObjectCheckEnabled     NO       YES

If Using NSDeallocateZombies=NO, then all objects will leak so be sure to turn it off when debugging memory leaks.

For more info on these varaiables see [NSDebug.h](http://theshadow.uw.hu/iPhoneSDKdoc/Foundation.framework/NSDebug.h.html)

For malloc debugging:

	MallocStackLogging
	MallocStackLoggingNoCompact
	MallocScribble
	MallocPreScribble
	MallocGuardEdges
	MallocDoNotProtectPrelude
	MallocDoNotProtectPostlude
	MallocCheckHeapStart
	MallocCheckHeapEach

If you see a message like:

	2009-10-15 13:02:24.746 Tests[38615:40b] *** -[Foo class]: message sent to deallocated instance 0x1c8e680

Re-run (in gdb) with `MallocStackLogging=YES` (or `MallocStackLoggingNoCompact=YES`), then if you run under gdb:

	(gdb) shell malloc_history 38615 0x1c8e680

	ALLOC 0x1a9ad10-0x1a9ad6f [size=96]: thread_a024a500 |start | main | UIApplicationMain | GSEventRun | GSEventRunModal | CFRunLoopRunInMode | CFRunLoopRunSpecific | __NSThreadPerformPerform | -[GHTestGroup _run:] | -[GHTest run] | +[GHTesting runTest:selector:withObject:exception:interval:] | -[Foo foo] | +[NSObject alloc] | +[NSObject allocWithZone:] | _internal_class_createInstance | _internal_class_createInstanceFromZone | calloc | malloc_zone_calloc 

Somewhere between runTest and NSObject alloc there may be an object that wasn't retained.

Also using `MallocScribble=YES` causes the malloc library to overwrite freed memory with a well-known value (0x55), and occasionally checks freed malloc blocks to make sure the memory has not been over-written overwritten written since it was cleared.

For more info on these variables see [MallocDebug](http://developer.apple.com/mac/library/documentation/Performance/Conceptual/ManagingMemory/Articles/MallocDebug.html)

For more info on malloc_history see [malloc_history](http://developer.apple.com/mac/library/documentation/Darwin/Reference/ManPages/man1/malloc_history.1.html)

## Command Line

To run the tests from the command line:

- Copy the [RunTests.sh](http://github.com/gabriel/gh-unit/tree/master/Scripts/RunTests.sh) file into your project directory (if you haven't already).
- In XCode:
  - To the `Tests` target, Add `New Build Phase` | `New Run Script Build Phase`
  - Enter `sh RunTests.sh` as the script. The path to `RunTests.sh` should be relative to the xcode project file (.xcodeproj)!
	- (Optional) Uncheck 'Show environment variables in build log'

From the command line, run the tests from xcodebuild (with the GHUNIT_CLI environment variable set) :

	// For mac app; This might seg fault in 10.6, in which case you should use `make test` via Makefile below
	GHUNIT_CLI=1 xcodebuild -target Tests -configuration Debug -sdk macosx10.5 build	
	
	// For iPhone app
	GHUNIT_CLI=1 xcodebuild -target Tests -configuration Debug -sdk iphonesimulator4.0 build

If you are wondering, the `RunTests.sh` script will only run the tests if the env variable GHUNIT_CLI is set. 
This is why this RunScript phase is ignored when running the test GUI. This is how we use a single Test target for both the GUI and command line testing.

This may seem strange that we run via xcodebuild with a RunScript phase in order to work on the command line, but otherwise we may not have
the environment settings or other XCode specific configuration right.

## Makefile

Follow the directions above for adding command line support.

Example Makefile's for Mac or iPhone apps:

- [Makefile](http://github.com/gabriel/gh-unit/tree/master/Project/Makefile) (for a Mac App)
- [Makefile](http://github.com/gabriel/gh-unit/tree/master/Project-IPhone/Makefile) (for an iPhone App)

The script will return a non-zero exit code on test failure.

To run the tests via the Makefile:

	make test

## Running a Test Case / Single Test

The `TEST` environment variable can be used to run a single test or test case.

	// Run all tests in GHSlowTest
	make test TEST="GHSlowTest"
	
	// Run the method testSlowA in GHSlowTest	
	make test TEST="GHSlowTest/testSlowA"

## Custom Test Case Classes

You can register additional classes at runtime; if you have your own. For example:

        [[GHTesting sharedInstance] registerClassName:@"MySpecialTestCase"];



## Test Macros
 
The following test macros are included. 

These macros are directly from: [GTMSenTestCase.h](http://code.google.com/p/google-toolbox-for-mac/source/browse/trunk/UnitTesting/GTMSenTestCase.h)
prefixed with GH so as not to conflict with the GTM macros if you are using those in your project.

The `description` argument appends extra information for when the assert fails; though most of the time you might leave it as nil.
 
	GHAssertNoErr(a1, description, ...)
	GHAssertErr(a1, a2, description, ...)
	GHAssertNotNULL(a1, description, ...)
	GHAssertNULL(a1, description, ...)
	GHAssertNotEquals(a1, a2, description, ...)
	GHAssertNotEqualObjects(a1, a2, desc, ...)
	GHAssertOperation(a1, a2, op, description, ...)
	GHAssertGreaterThan(a1, a2, description, ...)
	GHAssertGreaterThanOrEqual(a1, a2, description, ...)
	GHAssertLessThan(a1, a2, description, ...)
	GHAssertLessThanOrEqual(a1, a2, description, ...)
	GHAssertEqualStrings(a1, a2, description, ...)
	GHAssertNotEqualStrings(a1, a2, description, ...)
	GHAssertEqualCStrings(a1, a2, description, ...)
	GHAssertNotEqualCStrings(a1, a2, description, ...)
	GHAssertEqualObjects(a1, a2, description, ...)
	GHAssertEquals(a1, a2, description, ...)
	GHAbsoluteDifference(left,right) (MAX(left,right)-MIN(left,right))
	GHAssertEqualsWithAccuracy(a1, a2, accuracy, description, ...)
	GHFail(description, ...)
	GHAssertNil(a1, description, ...)
	GHAssertNotNil(a1, description, ...)
	GHAssertTrue(expr, description, ...)
	GHAssertTrueNoThrow(expr, description, ...)
	GHAssertFalse(expr, description, ...)
	GHAssertFalseNoThrow(expr, description, ...)
	GHAssertThrows(expr, description, ...)
	GHAssertThrowsSpecific(expr, specificException, description, ...)
	GHAssertThrowsSpecificNamed(expr, specificException, aName, description, ...)
	GHAssertNoThrow(expr, description, ...)
	GHAssertNoThrowSpecific(expr, specificException, description, ...)
	GHAssertNoThrowSpecificNamed(expr, specificException, aName, description, ...)

## Using an Alternate iPhone Application Delegate

If you want to use a custom application delegate in your test environment, you should subclass GHUnitIPhoneAppDelegate:

		@interface MyTestApplicationDelegate : GHUnitIPhoneAppDelegate { }
		@end

Then in GHUnitIPhoneTestMain.m:

		retVal = UIApplicationMain(argc, argv, nil, @"MyTestApplicationDelegate");

I am looking into removing this dependency but this will work in the meantime.

## Using SenTestingKit

You can also use GHUnit with SenTestCase, for example:

	#import <SenTestingKit/SenTestingKit.h>

	@interface MyTest : SenTestCase { }
	@end

	@implementation MyTest

	- (void)setUp {
		// Run before each test method
	}

	- (void)tearDown {
		// Run after each test method
	}

	- (void)testFoo {
		// Assert a is not NULL, with no custom error description
		STAssertNotNULL(a, nil);
	
		// Assert equal objects, add custom error description
		STAssertEqualObjects(a, b, @"Foo should be equal to: %@. Something bad happened", bar);
	}

	- (void)testBar {
		// Another test
	}

	@end

## Hudson

For integrating with Hudson (and generating JUnit XML) see HUDSON-HOWTO; Thank mikelaurence and precipice for their work on this!

## Building

There should be working Makefile's in Project and Project-IPhone directories.

## Notes

GHUnit was inspired by and uses parts of GTM (google-toolbox-for-mac) code, most from [UnitTesting](http://code.google.com/p/google-toolbox-for-mac/source/browse/trunk/UnitTesting/).


CHANGELOG.md
# Custom modifications
- Allow iPhone UI to handle stacked test groups
- Remove search bar

# Release 0.4.26
- Fixing LLVM/clang warnings (Whitney Young, nolanw)
- GHAssertNotEqualStrings will allow for nils (Rusty Zarse)
- Build warnings under 10.6 (MacOSX); (zykloid)
- Better error handling on JUnit XML results writing (zykloid)
- GHAsyncTestCase#runForInterval (Adapted from Robert Palmer, pauseForTimeout)

# Release 0.4.25
- Set DEPLOYMENT_POSTPROCESSING (MacOSX); So breakpointing doesn't warn about missing symbols

# Release 0.4.24
- Moved build settings into xcconfig (MacOSX)
- Striping linked build

# Release 0.4.21
- Moved build settings into xcconfig (iPhone)
- Flexible layouts; Works in iPad as universal app

# Release 0.4.20
- Fix armv6/armv7 device build setting

# Release 0.4.19
- Fix autorun env on iPhone
- Added re-run test (experimental!)
- Test log viewer (iPhone)
- Showing time in tests vs time running

# Release 0.4.18
- Fixing test stats on parallel running
- Adding reraiseException options (MacOSX)
- Adding env var support for reraise and autorun (see README)
- Smaller font size for test view (iPhone)
- Show filename/line number in trace on failure
- Show link to exception filename on failure (MacOSX)
- Fix bug where test trace/log doesn't update if selected before running

# Release 0.4.17
- Fixing disabled on new test bug
- Fixing bugs with All/Failed/Edit views not showing tests properly (MacOSX)

# Release 0.4.16
- Fixing hidden tests bug

# Release 0.4.15
- Text filter (MacOSX)
- Text filter now searches test case and test names (prefix)
- Failed filter (MacOSX/iPhone)
- Copy text in text view (MacOSX)
- Remember test state

# Release 0.4.14
- Fix window resizing when showing details
- Adding test for 0 found test cases

# Release 0.4.13
- Fixing framework build: Header error and 32/64 bit universal (MacOSX)
- Fixing SenTest macros not failing correctly
- Fixing persist of test enabled/disabled state
- Fixing SenTest macros

# Release 0.4.12
- Fixing compile warning in main (iPhone)

# Release 0.4.11
- Added value formatter (from http://github.com/JohannesRudolph); For better Assert error messages.
- Fixed deprecation warning (iPhone)
- Added default exception handler to give stack trace if triggered outside of GHUnit run

# Release 0.4.10
- Added Search Bar
- Added GHTestSuite#suiteWithPrefix:options

# Release 0.4.9
- Fix compile warning

# Release 0.4.8
- Fix bug with turning Parallel off not working
- Building 32/64 bit universal

# Release 0.4.7
- Removing redirect, was a bad idea; Test output goes to stderr, you can redirect stdout yourself

# Release 0.4.6
- Redirecting test output to file
- Test output does OK/FAIL
- Disabled tests appear gray (MacOSX)
- UI fixes

# Release 0.4.5 (2008-07-21)
- Including GHUnitIPhoneAppDelegate so you can subclass and interact with UIApplication delegate in tests

# Release 0.4.4 (2008-07-20)
- Ignore disable/cancelled tests in scroll (iPhone)
- Only start group test (notify) if we have tests to run

# Release 0.4.3 (2008-07-20)
- When running test on main thread should wait until finished
- Auto scrolls to middle instead of bottom (iPhone)

# Release 0.4.2 (2008-07-19)
- Fixing run warning

# Release 0.4.1 (2008-07-18)
- Option to use NSOperationQueue to manage tests runs
- Updated how test groups run
- Handling failure in setUpClass/tearDownClass
- Updated how shouldRunOnMainThread works
- Added Edit UI for Mac OSX tests
- Bug fixes and other refactoring

# Release 0.3.19 (2008-06-15)
- Fixed bug in Edit->Save crash (iPhone)
- Tweaking test text color (iPhone)
- Re-run crash
- Added reset to GHTest protocol
- Added testDidUpdate: to GHTestDelegate protocol
- On Edit->Save, triggers reset
- Added cancel to GHTest protocol
- Added cancelling, cancelled enums to test status
- Changed testDidFinish to testDidEnd (since test may be cancelled)

# Release 0.3.18 (2008-06-15)
- Adding Run button; By default tests do not automatically run on start
- Added AutoRun setting

# Release 0.3.17 (2008-06-09)
- Updating RunTests.sh

# Release 0.3.16 (2008-06-09)
- Rebuilding from 3.0 GM
- Setting debug variables in main directly instead of from setenv (which doesn't seem to work)

# Release 0.3.14 (2008-06-08)
- Creating separate iPhone 3.0 builds

# Release 0.3.12 (2008-05-25)
- Creating iPhone static library with device and simulator platforms

# Release 0.3.11 (2008-05-20)
- Fixing version number
- Creating separate version with CoreLocation linked

# Release 0.3.10 (2008-05-20)
- Fix namespace issue

# Release 0.3.9 (2008-05-19)

## 2008-05-19
- 3.0 compatibility fixes
- Added GHUITestCase
- Added shouldRunOnMainThread to test case, and if present and YES will run the tests on the main thread

## 2008-05-05
- (iPhone) Added select/deselect to iPhone test UI
- (iPhone) Fixed auto-scroll if you manually scroll (will stop auto-scrolling)

# Release 0.3.8 (2008-04-28)

## 2008-04-28
- Removed button enabled cell from Mac OS X view; Makes NSOutlineView really slow; Need to figure out how to do it right

# Release 0.3.7 (2008-04-26)

## 2008-04-20
- CLLocationManager mock
- Fixed afterDelay not using delay value
- Select/unselect (ignore) tests in Mac OSX view
- Added initWithTestSuite to GHTestApp for custom suites from test main

## 2008-04-16
- Adding ability to set run loops in async test case
- Adding more methods to NSURLConnection mock

# Release 0.3.6 (2008-04-13)

## 2008-04-13
- Adding swizzle methods for mocking
- Adding NSLocale mock
- Adding NSURLConnection, NSHTTPURLResponse mocks
- Fix bug with setUpClass/tearDownClass not working for single command line tests
- Setting Installation Directory to @rpath (Thanks chapados), so you can embed the framework with your app
- Sorting tests by class name (as well as method name)

# Release 0.3.4 (2009-04-11)

## 2008-04-11
- Added Doxygen support

## 2008-04-08
- Added GHAsyncTestCase for asynchronous tests (seems really complex :/, might have gone mental on it)

Classes/GHTest/GHTest.m
//
//  GHTest.m
//  GHKit
//
//  Created by Gabriel Handford on 1/18/09.
//  Copyright 2009. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person
//  obtaining a copy of this software and associated documentation
//  files (the "Software"), to deal in the Software without
//  restriction, including without limitation the rights to use,
//  copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the
//  Software is furnished to do so, subject to the following
//  conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
//  OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
//  OTHER DEALINGS IN THE SOFTWARE.
//

#import "GHTest.h"
#import "GHTest+JUnitXML.h"

#import "GHTesting.h"

@interface GHTest ()
- (void)_setLogWriter:(id<GHTestCaseLogWriter>)logWriter;
@end

NSString* NSStringFromGHTestStatus(GHTestStatus status) {
  switch(status) {
    case GHTestStatusNone: return NSLocalizedString(@"Waiting", nil);
    case GHTestStatusRunning: return NSLocalizedString(@"Running", nil);
    case GHTestStatusCancelling: return NSLocalizedString(@"Cancelling", nil);
    case GHTestStatusSucceeded: return NSLocalizedString(@"Succeeded", nil); 
    case GHTestStatusErrored: return NSLocalizedString(@"Errored", nil);
    case GHTestStatusCancelled: return NSLocalizedString(@"Cancelled", nil);
      
    default: return NSLocalizedString(@"Unknown", nil);
  }
}

GHTestStats GHTestStatsMake(NSInteger succeedCount, NSInteger failureCount, NSInteger cancelCount, NSInteger testCount) {
  GHTestStats stats;
  stats.succeedCount = succeedCount;
  stats.failureCount = failureCount; 
  stats.cancelCount = cancelCount;  
  stats.testCount = testCount;
  return stats;
}

const GHTestStats GHTestStatsEmpty = {0, 0, 0, 0};

NSString *NSStringFromGHTestStats(GHTestStats stats) {
  return [NSString stringWithFormat:@"%d/%d/%d/%d", stats.succeedCount, stats.failureCount, 
          stats.cancelCount, stats.testCount]; 
}

BOOL GHTestStatusIsRunning(GHTestStatus status) {
  return (status == GHTestStatusRunning || status == GHTestStatusCancelling);
}

BOOL GHTestStatusEnded(GHTestStatus status) {
  return (status == GHTestStatusSucceeded 
          || status == GHTestStatusErrored
          || status == GHTestStatusCancelled);
}

@implementation GHTest

@synthesize delegate=delegate_, target=target_, selector=selector_, name=name_, interval=interval_, 
exception=exception_, status=status_, log=log_, identifier=identifier_, disabled=disabled_, hidden=hidden_,
className=className_;

- (id)initWithIdentifier:(NSString *)identifier name:(NSString *)name {
  if ((self = [self init])) {
    identifier_ = [identifier retain];
    name_ = [name retain];
    className_ = [identifier retain];
    interval_ = -1;
    status_ = GHTestStatusNone;
  }
  return self;
}

- (id)initWithTarget:(id)target selector:(SEL)selector {
  NSString *name = NSStringFromSelector(selector);
  NSString *identifier = [NSString stringWithFormat:@"%@/%@", NSStringFromClass([target class]), name];
  if ((self = [self initWithIdentifier:identifier name:name])) {
    target_ = [target retain];
    selector_ = selector;
	className_ = [NSStringFromClass([target class]) retain];
  }
  return self;  
}

+ (id)testWithTarget:(id)target selector:(SEL)selector {
  return [[[self alloc] initWithTarget:target selector:selector] autorelease];
}

- (void)dealloc {
  [identifier_ release];
  [name_ release];
  [target_ release];
  [exception_ release];
  [log_ release]; 
  [super dealloc];
}

- (BOOL)isEqual:(id)test {
  return ((test == self) || 
          ([test conformsToProtocol:@protocol(GHTest)] && 
           [self.identifier isEqual:[test identifier]]));
}

- (NSString *)description {
  return [NSString stringWithFormat:@"%@ %@", self.identifier, [super description]];
}

- (GHTestStats)stats {
  switch(status_) {
    case GHTestStatusSucceeded: return GHTestStatsMake(1, 0, 0, 1);
    case GHTestStatusErrored: return GHTestStatsMake(0, 1, 0, 1);
    case GHTestStatusCancelled: return GHTestStatsMake(0, 0, 1, 1);
    default:
      return GHTestStatsMake(0, 0, 0, 1);
  }
}

- (void)reset {
  status_ = GHTestStatusNone;
  interval_ = 0;
  [exception_ release];
  exception_ = nil; 
  [delegate_ testDidUpdate:self source:self];
}

- (void)cancel {
  if (status_ == GHTestStatusRunning) {
    status_ = GHTestStatusCancelling;
    // TODO(gabe): Call cancel on target if available?    
  } else {
    status_ = GHTestStatusCancelled;
  }
  [delegate_ testDidUpdate:self source:self];
}

- (void)setDisabled:(BOOL)disabled {
  disabled_ = disabled;
  [delegate_ testDidUpdate:self source:self];
}

- (void)setHidden:(BOOL)hidden {
  hidden_ = hidden;
  [delegate_ testDidUpdate:self source:self];
}

- (NSInteger)disabledCount {
  return (disabled_ || hidden_ ? 1 : 0);
}

- (void)setException:(NSException *)exception {
  [exception retain];
  [exception_ release];
  exception_ = exception;
  status_ = GHTestStatusErrored;
  [delegate_ testDidUpdate:self source:self];
}

- (void)setUpClass {
  // Set up class
  @try {    
    if ([target_ respondsToSelector:@selector(setUpClass)]) {
      [target_ setUpClass];
    }
  } @catch(NSException *exception) {
    // If we fail in the setUpClass, then we will cancel all the child tests (below)
    exception_ = [exception retain];
    status_ = GHTestStatusErrored;
  }
}

- (void)tearDownClass {
  // Tear down class
  @try {
    if ([target_ respondsToSelector:@selector(tearDownClass)])    
      [target_ tearDownClass];
  } @catch(NSException *exception) {          
    exception_ = [exception retain];
    status_ = GHTestStatusErrored;
  }
}

- (void)run:(GHTestOptions)options {
  if (status_ == GHTestStatusCancelled || disabled_ || hidden_) return;
  
  if ((options & GHTestOptionForceSetUpTearDownClass) == GHTestOptionForceSetUpTearDownClass) {
    [self setUpClass];
    if (status_ == GHTestStatusErrored) return; 
  }
  
  status_ = GHTestStatusRunning;
  
  [delegate_ testDidStart:self source:self];
  
  [self _setLogWriter:self];

  BOOL reraiseExceptions = ((options & GHTestOptionReraiseExceptions) == GHTestOptionReraiseExceptions);
  [GHTesting runTestWithTarget:target_ selector:selector_ exception:&exception_ interval:&interval_ reraiseExceptions:reraiseExceptions];
  
  [self _setLogWriter:nil];

  if (exception_) {
    status_ = GHTestStatusErrored;
  }

  if (status_== GHTestStatusCancelling) {
    status_ = GHTestStatusCancelled;
  } else if (status_ == GHTestStatusRunning) {
    status_ = GHTestStatusSucceeded;
  }
  
  if ((options & GHTestOptionForceSetUpTearDownClass) == GHTestOptionForceSetUpTearDownClass)
    [self tearDownClass];
  
  [delegate_ testDidEnd:self source:self];
}

- (void)log:(NSString *)message testCase:(id)testCase {
  if (!log_) log_ = [[NSMutableArray array] retain];
  [log_ addObject:message];
  [delegate_ test:self didLog:message source:self];
}

#pragma mark Log Writer

- (void)_setLogWriter:(id<GHTestCaseLogWriter>)logWriter {
  if ([target_ respondsToSelector:@selector(setLogWriter:)])
    [target_ setLogWriter:logWriter];
} 

#pragma mark NSCoding

- (void)encodeWithCoder:(NSCoder *)coder {
  [coder encodeObject:identifier_ forKey:@"identifier"];
  [coder encodeBool:hidden_ forKey:@"hidden"];
  [coder encodeInteger:status_ forKey:@"status"];
  [coder encodeDouble:interval_ forKey:@"interval"];
}

- (id)initWithCoder:(NSCoder *)coder {
  GHTest *test = [self initWithIdentifier:[coder decodeObjectForKey:@"identifier"] name:nil];
  test.hidden = [coder decodeBoolForKey:@"hidden"];
  test.status = [coder decodeIntegerForKey:@"status"];
  test.interval = [coder decodeDoubleForKey:@"interval"];
  return test;
}

#pragma mark NSCopying

- (id)copyWithZone:(NSZone *)zone {
  if (!target_) [NSException raise:NSObjectNotAvailableException format:@"NSCopying unsupported for tests without target/selector pair"];
  return [[GHTest allocWithZone:zone] initWithTarget:target_ selector:selector_];
}

@end
